src/provider/provider_fixed_memory.c

```diff
@@ -0,0 +1,335 @@
+    coarse_params.cb.alloc = NULL;
+    coarse_params.cb.free = NULL; // not available for the fixed provider
+    coarse_params.cb.split = fixed_allocation_split_cb;
+    coarse_params.cb.merge = fixed_allocation_merge_cb;
+
+    coarse_t *coarse = NULL;
+    ret = coarse_new(&coarse_params, &coarse);
+    if (ret != UMF_RESULT_SUCCESS) {
+        LOG_ERR("coarse_new() failed");
+        goto err_free_fixed_provider;
+    }
+
+    fixed_provider->coarse = coarse;
+
+    fixed_provider->base = in_params->ptr;
+    fixed_provider->size = in_params->size;
+
+    // add the entire memory as a single block
+    ret = coarse_add_memory_fixed(coarse, fixed_provider->base,
+                                  fixed_provider->size);
+    if (ret != UMF_RESULT_SUCCESS) {
+        LOG_ERR("adding memory block failed");
+        goto err_coarse_delete;
+    }
+
+    *provider = fixed_provider;
+
+    return UMF_RESULT_SUCCESS;
+
+err_coarse_delete:
+    coarse_delete(fixed_provider->coarse);
+err_free_fixed_provider:
+    umf_ba_global_free(fixed_provider);
+    return ret;
+}
+
+static void fixed_finalize(void *provider) {
+    fixed_memory_provider_t *fixed_provider = provider;
+    coarse_delete(fixed_provider->coarse);
+    umf_ba_global_free(fixed_provider);
+}
+
+static umf_result_t fixed_alloc(void *provider, size_t size, size_t alignment,
+                                void **resultPtr) {
+    fixed_memory_provider_t *fixed_provider =
+        (fixed_memory_provider_t *)provider;
+    size_t page_size = utils_get_page_size();
+    if (alignment < page_size) {
+        alignment = page_size;
+    }
+    return coarse_alloc(fixed_provider->coarse, size, alignment, resultPtr);
+}
+
+static void fixed_get_last_native_error(void *provider, const char **ppMessage,
+                                        int32_t *pError) {
+    (void)provider; // unused
+
+    if (ppMessage == NULL || pError == NULL) {
+        assert(0);
+        return;
+    }
+
+    *pError = TLS_last_native_error.native_error;
+    if (TLS_last_native_error.errno_value == 0) {
+        *ppMessage = Native_error_str[*pError - UMF_FIXED_RESULT_SUCCESS];
+        return;
+    }
+
+    const char *msg;
+    size_t len;
+    size_t pos = 0;
+
+    msg = Native_error_str[*pError - UMF_FIXED_RESULT_SUCCESS];
+    len = strlen(msg);
+    memcpy(TLS_last_native_error.msg_buff + pos, msg, len + 1);
+    pos += len;
+
+    msg = ": ";
+    len = strlen(msg);
+    memcpy(TLS_last_native_error.msg_buff + pos, msg, len + 1);
+    pos += len;
+
+    utils_strerror(TLS_last_native_error.errno_value,
+                   TLS_last_native_error.msg_buff + pos, TLS_MSG_BUF_LEN - pos);
+
+    *ppMessage = TLS_last_native_error.msg_buff;
+}
+
+static umf_result_t fixed_get_recommended_page_size(void *provider, size_t size,
+                                                    size_t *page_size) {
+    (void)provider; // unused
+    (void)size;     // unused
+
+    *page_size = utils_get_page_size();
```

Comment:
`FIXED_PAGE_SIZE_2MB` !

Reply:
So what was `FIXED_PAGE_SIZE_2MB ` for?

Reply:
i forgot to remove it - fixed provider is based on dax one